src/Marten/Events/Daemon/Coordination/SingleTenantProjectionDistributor.cs

```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JasperFx.Core;
+using Marten.Storage;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Marten.Events.Daemon.Coordination;
+
+public class SingleTenantProjectionDistributor : IProjectionDistributor
+{
+    private readonly DocumentStore _store;
+    private readonly Cache<IMartenDatabase, AdvisoryLock> _locks;
+
+    public SingleTenantProjectionDistributor(DocumentStore store)
+    {
+        _store = store;
+
+        var logger = _store.Options.LogFactory?.CreateLogger<AdvisoryLock>() ??
+                     _store.Options.DotNetLogger ?? NullLogger<AdvisoryLock>.Instance;
+
+        _locks = new(db => new AdvisoryLock(db, logger));
+    }
+
+    public ValueTask<IReadOnlyList<IProjectionSet>> BuildDistributionAsync()
+    {
+        var database = _store.Storage.Database;
+        var projectionShards = _store.Options.Projections.AllShards();
+
+        var random = new Random();
+
+        IReadOnlyList<IProjectionSet> sets = projectionShards.Select(shard =>
+        {
+            // Make deterministic for each projection name
+            var lockId =
+                Math.Abs($"{_store.Options.EventGraph.DatabaseSchemaName}:{shard.Name.Identity}"
```

Comment:
**QUESTION:** Will (currently or in the future) shard name be different than `All`? Of what I see currently, even though there's a constructor allowing to pass different shard name, it's not used anywhere besides the default constructor with `IProjectionSource`.

Reply:
Sigh. This is a bit of future proofing that got into V4. It's from this: https://github.com/JasperFx/marten/issues/1715

We ran out of time and energy to do this in V4. At the time I though that we would provide "sharding" within some projections, and that's what that was meant for. Like being able to say things like arbitrarily dividing up streams into sets (based on a number using modulo operators maybe?), then running the same projection, but segmented in to multiple shards. Or being able to say "run this tenant in one shard, this other tenant in another shard, and lastly, all other tenants in this third shard"

I don't think we'll make any attempt to do anything w/ this in V7, but its' something I'd like to consider in the future. 

At this point, we enforce that every projection has a unique projection name. So every projection will have just the one "shard" for "All", but the hash will at least be unique. I incorporated the shard name into the hash as a future proofing thing. To sum that up, **every single shard will get a unique hash because it's built from the projection name, shard name, and schema name too**.

The inclusion of the schema name was to try to proof this against folks using separate stores in the same DB.

Reply:
Ok, understood; as we now have a single shard per projection, this will work. I was wondering if you want to include some additional sharding here, e.g., per tenant. Then it wouldn't be unique.